### PR TITLE
[ET-917] Add non-persistent cache to heavy calls

### PR DESCRIPTION
**Pull request by rafsuntaskin**
[ET-917]

We call various methods in Tribe_Tickets_Ticket_Object like $ticket->available() and $ticket->is_in_stock() in views. They always hit the DB unless they hit some meta cache layer. It would be very handy to be able to cache that to the object so subsequent calls aren't having to redo the logic. A simple reset method could be added to clear that cache too, or just simply get a new ticket object made from the provider as that would be unaffected by this work.

Before caching:

Screen Shot 2020-11-04 at 10 24 07 PM

After caching:

Screen Shot 2020-11-04 at 10 27 01 PM

[ET-917]: https://moderntribe.atlassian.net/browse/ET-917

### CODE REVIEW

File: src/Tribe/Tickets_Handler.php [hunk 1]
```diff
@@ -1195,6 +1195,14 @@ public function get_event_rsvp_tickets( $post = null ) {
 	 *             (allows zero but not `-1` for Unlimited). If oversold, will be corrected to zero.
 	 */
 	public function get_ticket_max_purchase( $ticket_id ) {
+
+		$cache = tribe( 'cache' );
+		$key   = __METHOD__ .'-'. $ticket_id;
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:1200]: * :no_entry_sign: Concat operator must be surrounded by a single space
* :no_entry_sign: Concat operator must be surrounded by a single space

<sub>hash: 5b227172fea519752362978e6712f7fe</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 2 error(s)

File: src/Tribe/Tickets_Handler.php (hunk 1 above)
Comment (bordoni) [reply src/Tribe/Tickets_Handler.php:1200]: Lets make sure we solve this.

Comment (sc0ttkclark) [reply src/Tribe/Tickets_Handler.php:1200]:
```suggestion
		$key   = __METHOD__ . '-' . $ticket_id;
```

File: src/Tribe/Ticket_Object.php [hunk 2]
```diff
@@ -654,6 +654,14 @@ public function remaining() {
 		 */
 		public function available() {
 			// if we aren't tracking stock, then always assume it is in stock or capacity is unlimited
+
+			$cache = tribe( 'cache' );
+			$key   = "ticket_available_{$this->ID}";
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:659]:
```suggestion
			$key   = __METHOD__ . '-' . $ticket_id;
```

**Review by sc0ttkclark — CHANGES_REQUESTED** [review]
Minor tweaks so far

File: src/Tribe/Tickets_Handler.php [hunk 3]
```diff
@@ -1195,6 +1195,14 @@ public function get_event_rsvp_tickets( $post = null ) {
 	 *             (allows zero but not `-1` for Unlimited). If oversold, will be corrected to zero.
 	 */
 	public function get_ticket_max_purchase( $ticket_id ) {
+
+		$cache = tribe( 'cache' );
+		$key = __METHOD__ . '-' . $ticket_id;
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:1200]: :warning: Equals sign not aligned with surrounding assignments; expected 3 spaces but found 1 space

<sub>hash: 50f72273d7e70de752a1c698ee373d95</sub>

**Review by tr1b0t — COMMENTED** [review]
:warning: 1 warning(s)

File: src/Tribe/Ticket_Object.php [hunk 4]
```diff
@@ -507,11 +507,20 @@ public function is_in_stock() {
 				return true;
 			}
 
-			$remaining = $this->inventory();
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:510]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 77985a88048f8626ef3a011374f088df</sub>

File: src/Tribe/Ticket_Object.php [hunk 5]
```diff
@@ -507,11 +507,20 @@ public function is_in_stock() {
...
+			$cache = tribe( 'cache' );
+			$key   = __METHOD__ . '-' . $this->ID;
 
-			$is_unlimited = $remaining === -1;
+			if ( isset( $cache[ $key ] ) ) {
+				return $cache[ $key ];
+			}
+
+			$remaining    = $this->inventory();
+			$is_unlimited = $remaining === - 1;
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:519]: :no_entry_sign: Use Yoda Condition checks, you must.

<sub>hash: e23a36edce53cfad224a513761989588</sub>

File: src/Tribe/Ticket_Object.php [hunk 6]
```diff
@@ -547,22 +556,32 @@ public function managing_stock( $manages_stock = null ) {
 		 * @return int
 		 */
 		public function inventory() {
+
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:560]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 9022193b6c14712fcc95b54a4cbb6a7e</sub>

File: src/Tribe/Ticket_Object.php [hunk 7]
```diff
@@ -684,6 +716,14 @@ public function capacity() {
 				return '';
 			}
 
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:719]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: ab21a1575853d7e6410a8f33f52ee436</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 1 error(s)
:warning: 3 warning(s)

File: src/Tribe/Ticket_Object.php (hunk 5 above)
Comment (sc0ttkclark) [reply src/Tribe/Ticket_Object.php:519]:
```suggestion
			$is_unlimited = - 1 === $remaining;
```

File: src/Tribe/Ticket_Object.php [hunk 8]
```diff
@@ -654,6 +674,14 @@ public function remaining() {
 		 */
 		public function available() {
 			// if we aren't tracking stock, then always assume it is in stock or capacity is unlimited
+
+			$cache = tribe( 'cache' );
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:678]:
```suggestion
			/** @var Tribe__Cache $cache */
			$cache = tribe( 'cache' );
```

File: src/Tribe/Tickets.php [hunk 9]
```diff
@@ -1464,6 +1475,13 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
 				return 0;
 			}
 
+			$cache = tribe( 'cache' );
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:1478]:
```suggestion
			/** @var Tribe__Cache $cache */
			$cache = tribe( 'cache' );
```

File: src/Tribe/Tickets.php [hunk 10]
```diff
@@ -1487,9 +1509,10 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
 		 * @return array
 		 */
 		public static function get_all_event_tickets( $post_id ) {
-			$cache_key = self::$cache_key_prefix . $post_id;
-			$cache = new Tribe__Cache();
-			$tickets = $cache->get( $cache_key );
+
+			$key     = __METHOD__ . '-' . $post_id;
+			$cache   = tribe( 'cache' );
+			$tickets = $cache[ $key ];
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:1515]:
```suggestion
			$key = __METHOD__ . '-' . $post_id;

			/** @var Tribe__Cache $cache */
			$cache   = tribe( 'cache' );
			$tickets = $cache[ $key ];
```

File: src/Tribe/Tickets.php [hunk 11]
```diff
@@ -1487,9 +1509,10 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
...
 
 			if ( is_array( $tickets ) ) {
 				return $tickets;
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:1518]:
```suggestion
			$key = __METHOD__ . '-' . $post_id;

			/** @var Tribe__Cache $cache */
			$cache   = tribe( 'cache' );

			if ( isset( $cache[ $key ] ) ) {
				return $cache[ $key ];
```

**Review by sc0ttkclark — CHANGES_REQUESTED** [review]
Some minor tweaks, but needs a reset ability

**Comment by sc0ttkclark** [comment]
Without reset ability, we could potentially have issues with automated testing too

File: src/Tribe/Ticket_Object.php [hunk 12]
```diff
@@ -507,11 +516,22 @@ public function is_in_stock() {
 				return true;
 			}
 
-			$remaining = $this->inventory();
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:519]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 4c4356cbdc40a2da8ee707314028000e</sub>

File: src/Tribe/Ticket_Object.php [hunk 13]
```diff
@@ -507,11 +516,22 @@ public function is_in_stock() {
...
+
+			if ( isset( $cache[ $key ] ) ) {
+				return $cache[ $key ];
+			}
+			// Storing the $key.
+			$this->cached_keys[] = $key;
 
-			$is_unlimited = $remaining === -1;
+			$remaining    = $this->inventory();
+			$is_unlimited = $remaining === - 1;
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:530]: :no_entry_sign: Use Yoda Condition checks, you must.

<sub>hash: c9ddc7cdd4efb78e0b6855de6ec877d4</sub>

File: src/Tribe/Ticket_Object.php [hunk 14]
```diff
@@ -547,22 +567,36 @@ public function managing_stock( $manages_stock = null ) {
 		 * @return int
 		 */
 		public function inventory() {
+
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:571]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: fd2868c3b3ae3d98ef4d6bf4889e9c78</sub>

File: src/Tribe/Ticket_Object.php [hunk 15]
```diff
@@ -684,6 +732,16 @@ public function capacity() {
 				return '';
 			}
 
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:735]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: bbda12915c3654ee0630326b3fd5eaaa</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 1 error(s)
:warning: 3 warning(s)

**Comment by rafsuntaskin** [comment]
> Without reset ability, we could potentially have issues with automated testing too

@sc0ttkclark 

Tribe_Cache doesn't remove non-persistent cache at this moment.

I have created a PR for that and the reset layer depends on that one: https://github.com/moderntribe/tribe-common/pull/1466

File: src/Tribe/Ticket_Object.php [hunk 16]
```diff
@@ -511,16 +511,29 @@ public function original_stock() {
...
 		 */
 		public function is_in_stock() {
-			// if we aren't tracking stock, then always assume it is in stock
+			// if we aren't tracking stock, then always assume it is in stock.
 			if ( ! $this->managing_stock() ) {
 				return true;
 			}
 
-			$remaining = $this->inventory();
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:519]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 1920a00c103a3d9d4340fb1a347edccd</sub>

File: src/Tribe/Ticket_Object.php [hunk 17]
```diff
@@ -556,22 +569,37 @@ public function managing_stock( $manages_stock = null ) {
 		 * @return int
 		 */
 		public function inventory() {
+
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:573]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 27c74cce08609b48a27c9bd307b193e9</sub>

File: src/Tribe/Ticket_Object.php [hunk 18]
```diff
@@ -662,7 +691,19 @@ public function remaining() {
 		 * @return int
 		 */
 		public function available() {
-			// if we aren't tracking stock, then always assume it is in stock or capacity is unlimited
+			// if we aren't tracking stock, then always assume it is in stock or capacity is unlimited.
+
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:696]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 21c5cf9f0d3aaea09954925f630abcc0</sub>

File: src/Tribe/Ticket_Object.php [hunk 19]
```diff
@@ -693,6 +738,18 @@ public function capacity() {
 				return '';
 			}
 
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Ticket_Object.php:741]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 2c8f2e4e71fc231bd9c2c7c09ad3df10</sub>

File: src/Tribe/Tickets.php [hunk 20]
```diff
@@ -680,6 +680,19 @@ public function post_has_tickets( $post ) {
 		 * @return Tribe__Tickets__Ticket_Object[] List of ticket objects.
 		 */
 		public function get_tickets( $post_id ) {
+
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:684]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: d02cd7634eb74f27d0b0c173785c20f6</sub>

File: src/Tribe/Tickets.php [hunk 21]
```diff
@@ -1491,6 +1506,18 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
 				return 0;
 			}
 
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:1509]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: f84976e22963452f8547eb15fe769322</sub>

File: src/Tribe/Tickets.php [hunk 22]
```diff
@@ -1514,12 +1545,18 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
 		 * @return array
 		 */
 		public static function get_all_event_tickets( $post_id ) {
-			$cache_key = self::$cache_key_prefix . $post_id;
-			$cache = new Tribe__Cache();
-			$tickets = $cache->get( $cache_key );
+			$key = __METHOD__ . '-' . $post_id;
+
+			/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:1550]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: f42f8934e574852394355c5fc120d75c</sub>

File: src/Tribe/Tickets_Handler.php [hunk 23]
```diff
@@ -1195,6 +1195,18 @@ public function get_event_rsvp_tickets( $post = null ) {
 	 *             (allows zero but not `-1` for Unlimited). If oversold, will be corrected to zero.
 	 */
 	public function get_ticket_max_purchase( $ticket_id ) {
+		/* @var Tribe__Cache $cache */
```
Comment (tr1b0t) [inline src/Tribe/Tickets_Handler.php:1198]: :warning: This comment is 50% valid code; is this commented out code?

<sub>hash: 00900324f8a837508b885d6c6df67c2d</sub>

**Review by tr1b0t — COMMENTED** [review]
:warning: 8 warning(s)

File: src/Tribe/Ticket_Object.php [hunk 24]
```diff
@@ -511,16 +511,29 @@ public function original_stock() {
 		 * @return boolean
 		 */
 		public function is_in_stock() {
-			// if we aren't tracking stock, then always assume it is in stock
+			// if we aren't tracking stock, then always assume it is in stock.
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:514]:
```suggestion
			// If we aren't tracking stock, then always assume it is in stock.
```

**Review by sc0ttkclark — APPROVED** [review]

**Review by sc0ttkclark — CHANGES_REQUESTED** [review]
Needs changelog

File: src/Tribe/Ticket_Object.php [hunk 25]
```diff
@@ -511,16 +511,29 @@ public function original_stock() {
...
-			// if we aren't tracking stock, then always assume it is in stock
+			// If we aren't tracking stock, then always assume it is in stock.
 			if ( ! $this->managing_stock() ) {
 				return true;
 			}
 
-			$remaining = $this->inventory();
+			/* @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:521]:
```suggestion
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $this->ID;
```

Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:521]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $this->ID;
```

File: src/Tribe/Ticket_Object.php [hunk 26]
```diff
@@ -556,22 +569,37 @@ public function managing_stock( $manages_stock = null ) {
 		 * @return int
 		 */
 		public function inventory() {
+
+			/* @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:575]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $this->ID;
```

File: src/Tribe/Ticket_Object.php [hunk 27]
```diff
@@ -662,7 +691,19 @@ public function remaining() {
 		 * @return int
 		 */
 		public function available() {
-			// if we aren't tracking stock, then always assume it is in stock or capacity is unlimited
+			// if we aren't tracking stock, then always assume it is in stock or capacity is unlimited.
+
+			/* @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:698]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $this->ID;
```

File: src/Tribe/Ticket_Object.php [hunk 28]
```diff
@@ -693,6 +738,18 @@ public function capacity() {
 				return '';
 			}
 
+			/* @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Ticket_Object.php:743]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $this->ID;
```

File: src/Tribe/Tickets.php [hunk 29]
```diff
@@ -680,6 +680,20 @@ public function post_has_tickets( $post ) {
 		 * @return Tribe__Tickets__Ticket_Object[] List of ticket objects.
 		 */
 		public function get_tickets( $post_id ) {
+
+			/* @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = $post_id;
+			$sub_key   = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:687]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $post_id;
```

File: src/Tribe/Tickets.php [hunk 30]
```diff
@@ -1491,6 +1507,19 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
 				return 0;
 			}
 
+			/* @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = $post_id;
+			$sub_key   = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:1513]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $post_id;
```

File: src/Tribe/Tickets.php [hunk 31]
```diff
@@ -1514,12 +1547,17 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
 		 * @return array
 		 */
 		public static function get_all_event_tickets( $post_id ) {
-			$cache_key = self::$cache_key_prefix . $post_id;
-			$cache = new Tribe__Cache();
-			$tickets = $cache->get( $cache_key );
+			/** @var Tribe__Cache $cache */
+			$cache     = tribe( 'cache' );
+			$cache_key = $post_id;
+			$sub_key   = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:1553]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_object_' . $post_id;
```

File: src/Tribe/Tickets_Handler.php [hunk 32]
```diff
@@ -1195,6 +1195,19 @@ public function get_event_rsvp_tickets( $post = null ) {
...
+		$cache     = tribe( 'cache' );
+		$cache_key = $ticket_id;
+		$sub_key   = __METHOD__;
```
Comment (sc0ttkclark) [inline src/Tribe/Tickets_Handler.php:1201]:
```suggestion
			$cache       = tribe( 'cache' );
			$cache_key   = __METHOD__;
			$cache_group = 'tribe_tickets_handler_' . $ticket_id;
```

File: src/Tribe/Tickets.php [hunk 33]
```diff
@@ -1514,12 +1537,13 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
...
 		 */
 		public static function get_all_event_tickets( $post_id ) {
-			$cache_key = self::$cache_key_prefix . $post_id;
-			$cache = new Tribe__Cache();
-			$tickets = $cache->get( $cache_key );
 
-			if ( is_array( $tickets ) ) {
-				return $tickets;
+			/** @var Tribe__Cache $cache */
+			$cache   = tribe( 'cache' );
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:1542]: :warning: Equals sign not aligned with surrounding assignments; expected 1 space but found 3 spaces

<sub>hash: 2b863865142fd14ad2eda6885f0f34a5</sub>

File: src/Tribe/Tickets.php [hunk 34]
```diff
@@ -1514,12 +1537,13 @@ public static function get_event_attendees_count( $post_id, $args = [] ) {
...
+			$key     = __METHOD__ . '-' . $post_id;
```
Comment (tr1b0t) [inline src/Tribe/Tickets.php:1543]: :warning: Equals sign not aligned with surrounding assignments; expected 3 spaces but found 5 spaces

<sub>hash: 3a5f5021fb47f5b1bbed8712e486d588</sub>

**Review by tr1b0t — COMMENTED** [review]
:warning: 2 warning(s)

**Comment by rafsuntaskin** [comment]
@sc0ttkclark we had to revert back to single key caching for now as the multidimensional array is not supported with the ArrayAccess structure.

For forcing a reset we need to reset the whole Tribe_Cache instance with a new instance when needed.

File: src/Tribe/Tickets.php (hunk 34 above)
Comment (sc0ttkclark) [inline src/Tribe/Tickets.php:1543]:
```suggestion
			/** @var Tribe__Cache $cache */
			$cache = tribe( 'cache' );
			$key   = __METHOD__ . '-' . $post_id;
```

**Review by sc0ttkclark — DISMISSED** [review]
Approved for QA in branch

**Review by sc0ttkclark — DISMISSED** [review]

**Review by sc0ttkclark — APPROVED** [review]

**Pull request merged**